### PR TITLE
[DEV-10253] Parallel test runs to speed up continuous integration

### PR DESCRIPTION
**Pull request by kwhickey**
**Description:**
Reduce time it takes for CI to run the full test suite. 
Allow developers also to gain parallel test performance locally simply by calling `make tests`

**Technical details:**
- Gained speedups in THREE approaches
    1. Use `pytest-xdist` to run parallel pytest sessions on multiple underlying processes
    2. Employ matrixed builds in Travis CI to take on different subsets of unit and integration tests to divide the workloads up
    3. Separate CI steps into Travis CI `stages` to factor out redundant steps like `pip install` and `docker build` and allow static code analysis to run checks in parallel
- Refactored tests for concurrent process issues and race conditions
- Added a lot of `_fill_optional=True` to baker test models to fix tests
    - Found that if a django `call_command(...)` call was made to a command that used a `.sql` file with a `COMMIT` statement, it was causing the test model data to be committed, and constraint to be evaluated, which would fail if required fields were missing values.
- Refactored fixtures to setup separate tests databases per pytest-xdist worker 
- In `usaspending_api/tests/integration/test_setup_of_spark_dependencies.py`, added a singular test to be run before concurrent spark tests, so that we are not writing dependencies to the same location concurrently and failing due to a file write race condition
- Mocking connections to the download database, so it connects back to the main test database (the one specifically setup for an instance of a parallel test session worker) during tests
- Moving all improperly categorized automated integration tests in to their proper `/tests/integration/` folder

See [example passing Travis CI build](https://app.travis-ci.com/github/fedspendingtransparency/usaspending-api/builds/266321759) with new stages and jobs:
image

**TODO**
_Items that may still be left to do_
1. [x] Review if code coverage files created separately by each Travis matrixed build are being appended/combined to yield an _overall_ code coverage value by Code Climate. If not, figure out how to accumulate coverage from each matrixed build.
    - After a test run, that passed, it seems to stick with the first build run's code coverage (in that case 28% only). Needs fixing
    - See [Code Climate with Travis CI parallel tests](https://docs.codeclimate.com/docs/travis-ci-test-coverage#example-2) example
        1. At the end of each build job, it copies that job's partial coverage up to a external data store (s3 in this case)
        2. Then copies all files from that store (for all jobs finished to-date) down locally
        4. Then sums all
        5. Then updates code climate with the sum as of that job's completion
        6. Last job that finishes should be publishing the overall sum
     - May be able to avoid S3 and see if [Travis CI Workspaces (beta)](https://docs.travis-ci.com/user/using-workspaces/) would work instead: 

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
7. [NA] API documentation updated
8. [ ] Necessary PR reviewers:
    - [ ] Backend
    - [NA] Frontend <OPTIONAL>
    - [NA] Operations <OPTIONAL>
    - [NA] Domain Expert <OPTIONAL>
9. [x] Matview impact assessment completed
10. [x] Frontend impact assessment completed
11. [x] Data validation completed
12. [NA] Appropriate Operations ticket(s) created
13. [x] Jira Ticket [DEV-10253](https://federal-spending-transparency.atlassian.net/browse/DEV-10253):
    - [x] Link to this Pull-Request
    - [NA] Performance evaluation of affected (API | Script | Download)
    - [x] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -45,6 +45,7 @@ BROKER_DB_PASSWORD=root
 # ==== [Elasticsearch] ====
 # Where to connect to elasticsearch.
 # All values of ES_* must match what is in ES_HOSTNAME if BOTH are given
+# NOTE: DO NOT include trainling slash on ES_HOSTNAME url
```
Comment (dporth-frb) [inline .env.template:48]: Typo in "trainling"

**Review by dporth-frb — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 2]
```diff
@@ -29,56 +31,187 @@ env:
...
+    workspaces:
+      create:
+        name: pip
+        paths:
+          - $HOME/virtualenv/
+          - $TRAVIS_BUILD_DIR/usaspending_api.egg_info/
+    before_install: ""  # override default to no-op
+    install:
+    - travis_retry pip install setuptools==60.8.2
+    - travis_retry pip install .[dev]  # TODO: need to also capture usaspending_api.egg_info?
```
Comment (dporth-frb) [inline .travis.yml:59]: Do you still want this TODO in here?

File: requirements/requirements-dev.txt [hunk 3]
```diff
@@ -3,10 +3,13 @@ click==8.0.4
...
-model-bakery==1.5.*
+importlib-metadata<5  # need to pin <5 while flake8 is on 3.8.4
+mock==5.1.*
+model-bakery==1.13.*
 pre-commit==1.20.0
 pyspark==3.2.1
-pytest-cov==2.10.1
-pytest-django==4.2.0
-pytest==6.2.*
+pytest==7.4.*
```
Comment (dporth-frb) [inline requirements/requirements-dev.txt:11]: We have a requirements-test.txt now. Wouldn't it be better to have these test libraries added to that and that requirements-test.txt used?

**Review by dporth-frb — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 2 above)
Comment (kwhickey) [reply .travis.yml]: Done.

File: usaspending_api/conftest.py [hunk 4]
```diff
@@ -43,6 +58,60 @@
 baker.generators.add("usaspending_api.common.custom_django_fields.BooleanFieldWithDefault", lambda: False)
 
 
+def pytest_collection_modifyitems(session: pytest.Session, config: pytest.Config, items: List[pytest.Item]) -> None:
```
Comment (dporth-frb) [inline usaspending_api/conftest.py:61]: You should document the parameters here with a description.

**Review by dporth-frb — CHANGES_REQUESTED** [review]

File: usaspending_api/common/helpers/sql_helpers.py [hunk 5]
```diff
@@ -19,11 +19,11 @@ def build_dsn_string(db_settings):
...
 
-def get_database_dsn_string():
-    if DEFAULT_DB_ALIAS in settings.DATABASES:
-        return build_dsn_string(settings.DATABASES[DEFAULT_DB_ALIAS])
+def get_database_dsn_string(db_alias=DEFAULT_DB_ALIAS):
+    if db_alias in settings.DATABASES:
+        return build_dsn_string(settings.DATABASES[db_alias])
     else:
-        raise Exception("No valid database connection is configured")
+        raise Exception(f'No valid database connection is configured with alias "{db_alias}"')
```
Comment (dporth-frb) [inline usaspending_api/common/helpers/sql_helpers.py:26]: What if db alias is not provided. Do we really want to output a message saying "No database connection configured with alias None".

Perhaps you should output a message saying "db alias must be provided and cannot be None"

File: usaspending_api/common/helpers/generic_helper.py [hunk 6]
```diff
@@ -113,12 +108,27 @@ def convert_matview_to_view(matview_sql):
...
+    temp_sql_files = [
+        {"name": key, "sql_file": matview_dir / val["sql_filename"]} for key, val in MATERIALIZED_VIEWS.items()
+    ]
+    temp_sql_files += [
+        {"name": key, "sql_file": matview_dir / val["sql_filename"]} for key, val in CHUNKED_MATERIALIZED_VIEWS.items()
+    ]
+    return temp_sql_files
+
+
+def generate_matviews(materialized_views_as_traditional_views=False, parallel_worker_id: str = None):
```
Comment (dporth-frb) [inline usaspending_api/common/helpers/generic_helper.py:121]: You should document the new parameter here, `parallel_worker_id`

File: usaspending_api/common/helpers/generic_helper.py [hunk 7]
```diff
@@ -113,12 +108,27 @@ def convert_matview_to_view(matview_sql):
     return sql
 
 
-def generate_matviews(materialized_views_as_traditional_views=False):
+def get_temp_matview_sql_files_dict(matview_dir: Path):
```
Comment (dporth-frb) [inline usaspending_api/common/helpers/generic_helper.py:111]: You should add a pydoc here.

File: usaspending_api/common/helpers/generic_helper.py [hunk 8]
```diff
@@ -113,12 +108,27 @@ def convert_matview_to_view(matview_sql):
...
+
+def generate_matviews(materialized_views_as_traditional_views=False, parallel_worker_id: str = None):
     with connection.cursor() as cursor:
         cursor.execute(CREATE_READONLY_SQL)
         cursor.execute(DEPENDENCY_FILEPATH.read_text())
-        subprocess.call(f"python3 {MATVIEW_GENERATOR_FILE} --dest {DEFAULT_MATIVEW_DIR} --quiet", shell=True)
-        for matview_sql_lookup in TEMP_SQL_FILES:
+
+        matview_dir = DEFAULT_MATIVEW_DIR
+        if parallel_worker_id:
```
Comment (dporth-frb) [inline usaspending_api/common/helpers/generic_helper.py:127]: Can we make this if statement less implicit? Change it to `if parallel_worker_id is not None`

Comment (dporth-frb) [reply usaspending_api/common/helpers/generic_helper.py:127]: As it stands its more implicit than it needs to be.

File: usaspending_api/tests/conftest_spark.py [hunk 9]
```diff
@@ -48,9 +59,10 @@
 
 
 @fixture(scope="session")
-def s3_unittest_data_bucket_setup_and_teardown():
+def s3_unittest_data_bucket_setup_and_teardown(worker_id):
```
Comment (dporth-frb) [inline usaspending_api/tests/conftest_spark.py:62]: You should duck type this new parameter and describe it in the pydoc.

File: usaspending_api/tests/conftest_spark.py [hunk 10]
```diff
@@ -221,3 +235,1043 @@ def delta_lake_unittest_schema(spark: SparkSession, hive_unittest_metastore_db):
     yield DELTA_LAKE_UNITTEST_SCHEMA_NAME
 
     # The dependent hive_unittest_metastore_db fixture will take care of cleaning up this schema post-test
+
+
+@fixture
+def populate_broker_data(broker_server_dblink_setup):
```
Comment (dporth-frb) [inline usaspending_api/tests/conftest_spark.py:241]: pydoc on this function

File: usaspending_api/common/tests/unit/test_sqs_work_dispatcher.py [hunk 11]
```diff
@@ -41,1509 +44,1597 @@ def _patch_get_sqs_queue(fake_sqs_queue, monkeypatch):
     assert should_be_fake.url.split("/")[-1] == UNITTEST_FAKE_QUEUE_NAME
 
 
-@pytest.mark.usefixtures("_patch_get_sqs_queue")
-class SQSWorkDispatcherTests(TestCase):
-    def test_dispatch_with_default_numeric_message_body_succeeds(self):
-        """SQSWorkDispatcher can execute work on a numeric message body successfully
+def test_fake_sqs_queue_fixture(fake_sqs_queue):
```
Comment (dporth-frb) [inline usaspending_api/common/tests/unit/test_sqs_work_dispatcher.py:47]: pydoc?

File: Makefile [hunk 12]
```diff
@@ -107,9 +117,22 @@ print('Hadoop ' + spark.sparkContext._gateway.jvm.org.apache.hadoop.util.Version
...
+.PHONY: test-dbs
+createdb :=  #unset it
+test-dbs:  ## Trigger the setup of multiple test DBs that can be reused with pytest --numprocesses. Add createdb=true to force (re-)creation of Test DBs rather than reuse.
+	pytest ${if ${createdb},--create-db,} --reuse-db --numprocesses=auto --no-cov --disable-warnings -rP -vvv --capture=no --log-cli-level=WARNING --show-capture=log 2> /dev/null 'usaspending_api/tests/integration/test_setup_of_test_dbs.py::test_trigger_test_db_setup'
+
+.PHONY: test-spark-deps
+test-spark-deps:  ## Trigger a singular test in one pytest session that does nothing but cause Maven dependencies to be downloaded and cached through Ivy; reduces contention when parallel spark builds need the depdencies
+	pytest --no-cov --disable-warnings -r=fEs --verbosity=3 'usaspending_api/tests/integration/test_setup_of_spark_dependencies.py::test_preload_spark_jars'
+
 .PHONY: tests
```
Comment (dporth-frb) [inline Makefile:129]: Why doesn't this service reference the requirements-dev.txt service?

Comment (kwhickey) [reply Makefile:130]: Added `local-dev-setup` as a dependency, which itself calls `requiremets-dev`. 

It depends how far we want to go with this useful shortcut command
- a) just call it and get right to tests, assuming the developer has setup all prerequisites
- b) call it and make sure it sets up all the things (virtual env, pip installs, etc.) that are needed to successfully run tests

We can try it with the dependent setup baked in, but it does add a bit of lag to getting the tests started. May be good for first-time developers on the team but experienced developers may not want that lag in their workflow. We'll see what kind of feedback we get.

File: usaspending_api/common/tests/unit/test_sqs_work_dispatcher.py (hunk 11 above)
Comment (kwhickey) [reply usaspending_api/common/tests/unit/test_sqs_work_dispatcher.py:47]: Done

File: usaspending_api/tests/conftest_spark.py (hunk 10 above)
Comment (kwhickey) [reply usaspending_api/tests/conftest_spark.py:249]: Done

File: usaspending_api/tests/conftest_spark.py (hunk 9 above)
Comment (kwhickey) [reply usaspending_api/tests/conftest_spark.py]: Done

File: usaspending_api/common/helpers/generic_helper.py (hunk 8 above)
Comment (kwhickey) [reply usaspending_api/common/helpers/generic_helper.py:144]: It covers two scenarios, if the param is `None` or if the param comes in as empty string. So while slightly implicit it is comprehensive and less verbose. It's a common idiom in python for checking a string (or e.g. list) has value to it, so I think worth leaving as is.

File: usaspending_api/common/helpers/generic_helper.py (hunk 7 above)
Comment (kwhickey) [reply usaspending_api/common/helpers/generic_helper.py]: Done

File: usaspending_api/common/helpers/generic_helper.py (hunk 6 above)
Comment (kwhickey) [reply usaspending_api/common/helpers/generic_helper.py]: Done

File: usaspending_api/common/helpers/sql_helpers.py (hunk 5 above)
Comment (kwhickey) [reply usaspending_api/common/helpers/sql_helpers.py:28]: Done

File: usaspending_api/conftest.py (hunk 4 above)
Comment (kwhickey) [reply usaspending_api/conftest.py:61]: Done

**Review by dporth-frb — DISMISSED** [review]

**Review by dporth-frb — APPROVED** [review]

**Review by dporth-frb — APPROVED** [review]

**Pull request merged**